File: backend/src/services/recommendations/books/books.py
#
# Copyright GETOUT SAS - All Rights Reserved
# Unauthorized copying of this file, via any medium is strictly prohibited
# Proprietary and confidential
# Wrote by Julien Letoux <julien.letoux@epitech.eu>
#

import sys
import json

from getParameters import getParameters
from getBooksPool import getBooksPool
from applyWeightsAndScore import applyWeightsAndScore
from filterBooksPool import filterBooksPool


def recommendBooks(account: dict) -> list:
    parameters = getParameters(account)
    booksPool = getBooksPool(parameters)
    booksPool = filterBooksPool(parameters, booksPool)
    recommendations = applyWeightsAndScore(parameters, booksPool)
    return recommendations


def main():
    try:
        # Ensure the JSON string is properly formatted with double quotes
        if len(sys.argv) > 1:
            externalSessionAccount = json.loads(sys.argv[1])
            result = recommendBooks(externalSessionAccount)
            print(json.dumps(result[0:5]))
        else:
            print("No input provided. Please pass a JSON string as the first argument.")
    except json.JSONDecodeError as e:
        print(f"Error decoding JSON from input: {e}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")


if __name__ == "__main__":
    main()

File: backend/src/services/recommendationsCaching/books.ts
/*
** Copyright GETOUT SAS - All Rights Reserved
** Unauthorized copying of this file, via any medium is strictly prohibited
** Proprietary and confidential
** Wrote by Alexandre Chetrit <chetrit.pro@hotmail.com>
*/

import { type UUID } from 'crypto'

import { type BookResponse } from '@services/googlebooks/getBook'
import logger from '@services/middlewares/logging'
import { CacheError } from '@services/utils/customErrors'
import { getRedisClient } from '@services/utils/redisClient'

import {
  type BookRecommandationAlgorithmResonse,
  generateBooksRecommendations,
  retreiveFullRecommendationsFromIds
} from '@models/account/recommendBooks'

const redisClient = getRedisClient()

export async function getCurrentRecommendations (userId: UUID): Promise<BookResponse [] | null> {
  return await redisClient.get(`recommendation:books:current:${userId}`).then((data) => {
    return data != null ? JSON.parse(data) as BookResponse[] : null
  }).catch((err) => {
    throw new CacheError(`Failed to get current recommendations for user ${userId}: ${err}`)
  })
}

export async function getNextRecommendations (userId: UUID): Promise<BookResponse [] | null> {
  return await redisClient.get(`recommendation:books:next:${userId}`).then((data) => {
    return data != null ? JSON.parse(data) as BookResponse[] : null
  }).catch((err) => {
    console.error(`Failed to get next recommendations for user ${userId}: ${err}`)
    throw err
  })
}

export async function preloadFirstRecommendations (userId: UUID): Promise<void> {
  await generateBooksRecommendations(userId).then(async (recommendations: BookRecommandationAlgorithmResonse []) => {
    return await retreiveFullRecommendationsFromIds(recommendations.map((recommendation) => recommendation.id))
  }).then(async (fullRecommendations: BookResponse []) => {
    await setCurrentRecommendations(userId, fullRecommendations)
  }).catch((err) => {
    throw new CacheError(`Failed to preload first recommendations for user ${userId}: ${err}`)
  })
}

export async function preloadNextRecommendations (userId: UUID): Promise<void> {
  await generateBooksRecommendations(userId).then(async (recommendations: BookRecommandationAlgorithmResonse []) => {
    return await retreiveFullRecommendationsFromIds(recommendations.map((recommendation) => recommendation.id))
  }).then(async (fullRecommendations: BookResponse []) => {
    await setNextRecommendations(userId, fullRecommendations)
  }).catch((err) => {
    throw new CacheError(`Failed to preload next recommendations for user (${userId}: ${err})`)
  })
}

export async function setCurrentRecommendations (
  userId: UUID,
  data: BookResponse [],
  ttl: number = 86400
): Promise<void> {
  await redisClient.set(
    `recommendation:books:current:${userId}`,
    JSON.stringify(data), { EX: ttl }
  ).then(() => {
    logger.debug('Current books recommendations have been set in the cache.')
  }).catch((err) => {
    throw new CacheError(`Failed to set current recommendations for user ${userId}: ${err}`)
  })
}

export async function setNextRecommendations (
  userId: UUID,
  data: BookResponse [],
  ttl: number = 86400
): Promise<void> {
  await redisClient.set(
    `recommendation:books:next:${userId}`,
    JSON.stringify(data), { EX: ttl }
  ).then(() => {
    logger.debug('Next books recommendations have been set in the cache.')
  }).catch((err) => {
    throw new CacheError(`Failed to set next recommendations for user ${userId}: ${err}`)
  })
}
